File: backend/app/services/product_service.py
from typing import List, Optional
from bson import ObjectId
from ..db.mongodb import get_database
from ..models.product import ProductCreate, ProductUpdate, ProductInDB
from datetime import datetime

class ProductService:
    collection_name = "products"

    @staticmethod
    async def get_products() -> List[ProductInDB]:
        db = await get_database()
        products = []
        async for product in db.products.find():
            product["_id"] = str(product["_id"])
            products.append(ProductInDB(**product))
        return products

    @staticmethod
    async def get_product(product_id: str) -> Optional[ProductInDB]:
        db = await get_database()
        if not ObjectId.is_valid(product_id):
            return None
        product = await db.products.find_one({"_id": ObjectId(product_id)})
        if product:
            product["_id"] = str(product["_id"])
            return ProductInDB(**product)
        return None

    @staticmethod
    async def create_product(product: ProductCreate) -> ProductInDB:
        db = await get_database()
        product_dict = product.model_dump()
        product_dict["created_at"] = datetime.utcnow()
        product_dict["updated_at"] = datetime.utcnow()
        
        result = await db.products.insert_one(product_dict)
        created_product = await db.products.find_one({"_id": result.inserted_id})
        created_product["_id"] = str(created_product["_id"])
        return ProductInDB(**created_product)

    @staticmethod
    async def update_product(product_id: str, product: ProductUpdate) -> Optional[ProductInDB]:
        db = await get_database()
        if not ObjectId.is_valid(product_id):
            return None
            
        update_data = product.model_dump(exclude_unset=True)
        if update_data:
            update_data["updated_at"] = datetime.utcnow()
            await db.products.update_one(
                {"_id": ObjectId(product_id)},
                {"$set": update_data}
            )
            
        updated_product = await db.products.find_one({"_id": ObjectId(product_id)})
        if updated_product:
            updated_product["_id"] = str(updated_product["_id"])
            return ProductInDB(**updated_product)
        return None

    @staticmethod
    async def delete_product(product_id: str) -> bool:
        db = await get_database()
        if not ObjectId.is_valid(product_id):
            return False
            
        result = await db.products.delete_one({"_id": ObjectId(product_id)})
        return result.deleted_count > 0 

File: src/lib/services/product-service.ts
import { Product, ProductCreate, ProductUpdate } from '@/types/product';
import { connectToDatabase } from '@/lib/mongodb';
import { ObjectId } from 'mongodb';

interface ProductDocument extends Omit<Product, 'id'> {
  _id: ObjectId;
}

export class ProductService {
  static async getProducts(): Promise<Product[]> {
    const db = await connectToDatabase();
    const products = await db.collection<ProductDocument>('products').find().toArray();
    return products.map(product => ({
      ...product,
      id: product._id.toString(),
    }));
  }

  static async getProduct(id: string): Promise<Product | null> {
    const db = await connectToDatabase();
    const product = await db.collection<ProductDocument>('products').findOne({ _id: new ObjectId(id) });
    if (!product) return null;
    return {
      ...product,
      id: product._id.toString(),
    };
  }

  static async createProduct(product: ProductCreate): Promise<Product> {
    const db = await connectToDatabase();
    const now = new Date().toISOString();
    const newProduct = {
      ...product,
      created_at: now,
      updated_at: now,
    };
    const result = await db.collection('products').insertOne(newProduct);
    return {
      ...newProduct,
      id: result.insertedId.toString(),
    };
  }

  static async updateProduct(id: string, product: ProductUpdate): Promise<Product | null> {
    const db = await connectToDatabase();
    const now = new Date().toISOString();
    const { _id, ...updateData } = product as any; // Remove _id if present
    const update = {
      ...updateData,
      updated_at: now,
    };
    const result = await db.collection<ProductDocument>('products').findOneAndUpdate(
      { _id: new ObjectId(id) },
      { $set: update },
      { returnDocument: 'after' }
    );
    if (!result) return null;
    return {
      ...result,
      id: result._id.toString(),
    };
  }

  static async deleteProduct(id: string): Promise<boolean> {
    const db = await connectToDatabase();
    const result = await db.collection('products').deleteOne({ _id: new ObjectId(id) });
    return result.deletedCount > 0;
  }
} 